sys/kern/vfs_syscalls.c

```diff
@@ -87,19 +89,42 @@ static int vfs_create(proc_t *p, int fdat, char *pathname, int flags, int mode,
   return error;
 }
 
+static int vfs_check_open(vnode_t *v, int flags, cred_t *cred) {
+  mode_t accmode = 0;
+  switch (flags & O_ACCMODE) {
+    case O_RDONLY:
+      accmode = VREAD;
+      break;
+    case O_WRONLY:
+      accmode = VWRITE;
+      break;
+    case O_RDWR:
+      accmode = VREAD | VWRITE;
+      break;
+  }
+  if (flags & O_TRUNC)
+    accmode |= VWRITE;
+
+  return VOP_ACCESS(v, accmode, cred);
+}
+
 static int vfs_open(proc_t *p, file_t *f, int fdat, char *pathname, int flags,
                     int mode) {
   vnode_t *v;
-  int error;
+  int error, created = 0;
 
   if (flags & O_CREAT) {
-    if ((error = vfs_create(p, fdat, pathname, flags, mode, &v)))
+    if ((error = vfs_create(p, fdat, pathname, flags, mode, &v, &created)))
       return error;
   } else {
     if ((error = vfs_namelookupat(p, fdat, VNR_FOLLOW, pathname, &v)))
       return error;
   }
 
+  if (!created)
+    if ((error = vfs_check_open(v, flags, &p->p_cred)))
+      return error;
+
```

Comment:
Just making sure I understand it:
This is an optimization, right? We could always do `vfs_check_open()`, but it will always succeed if we created the file. So we don't _need_ the `created` parameter in `vfs_create()`, correct?

Reply:
No. We don't call `vfs_check_open()` when we created a file. That means if you created a file and you won't have access to it you can still write or read it unless you close it in your program. It is done like that in BSD and in Linux.

Reply:
I see. Thanks!